### PR TITLE
[RDY] [build] add NVIM_BUILD_STATIC build static linked

### PR DESCRIPTION
**Pull request by zhaozg**
- [x] add NVIM_BUILD_STATIC support.
- [x] support static link iconv.
- [x] support static link intl.
- [x] support build winpty from source and static link
- [x] support static link jemalloc. disabled on windows.
- [x] support static link unibilium.
- [x] lint and style ok
- [x] static link on windows works fine,but test fail unrelatived

`cmake  -DNVIM_BUILD_STATIC=ON` will enable static build, default dynamic.

If not found static library, will use dyanmic library.

### CODE REVIEW

**Comment by zhaozg** [comment]
Replace https://github.com/neovim/neovim/pull/7884

**Comment by zhaozg** [comment]
After merge, I have more PR.

**Comment by zhaozg** [comment]
ping @justinmk , any advice?

File: cmake/FindJeMalloc.cmake [hunk 1]
```diff
@@ -23,6 +23,10 @@ find_path(JEMALLOC_INCLUDE_DIR jemalloc/jemalloc.h
           PATHS ${PC_JEMALLOC_INCLUDEDIR} ${PC_JEMALLOC_INCLUDE_DIRS}
           ${LIMIT_SEARCH})
 
+if(WIN32)
+  set(JEMALLOC_USE_STATIC ON)
```
Comment (justinmk) [inline cmake/FindJeMalloc.cmake]: We don't use jemalloc at all, on Windows.

File: cmake/FindUnibilium.cmake [hunk 2]
```diff
@@ -23,6 +23,10 @@ find_path(UNIBILIUM_INCLUDE_DIR unibilium.h
           PATHS ${PC_UNIBILIUM_INCLUDEDIR} ${PC_UNIBILIUM_INCLUDE_DIRS}
           ${LIMIT_SEARCH})
 
+if(WIN32)
+  set(UNIBILIUM_USE_STATIC ON)
```
Comment (justinmk) [inline cmake/FindUnibilium.cmake]: Why force it always? Instead, perhaps a new option like `NVIM_BUILD_STATIC`. Currently it will only work for Windows, but we may later add support for static builds on unix.

File: third-party/CMakeLists.txt [hunk 3]
```diff
@@ -204,18 +213,6 @@ if(WIN32)
   elseif(TARGET_ARCH STREQUAL "X86")
     set(TARGET_ARCH ia32)
   endif()
-
-  GetBinaryDep(TARGET winpty
```
Comment (justinmk) [inline third-party/CMakeLists.txt]: What's the point of building winpty instead of using the pre-built dll?

Comment (zhaozg) [reply third-party/CMakeLists.txt]: Let nvim only depends windows OS dlls, and keep binary exe independ.

Any reaaon not do that.

File: cmake/FindJeMalloc.cmake (hunk 1 above)
Comment (zhaozg) [reply cmake/FindJeMalloc.cmake]: If mingw64 with jemalloc installed, Cmake will found itand do link. So it not always not use.

File: cmake/FindUnibilium.cmake (hunk 2 above)
Comment (zhaozg) [reply cmake/FindUnibilium.cmake]: Uniblium in mingw64 only static libs. To declare use static lib for disambiguation

File: third-party/CMakeLists.txt (hunk 3 above)
Comment (zhaozg) [reply third-party/CMakeLists.txt]: and winpty-adapter.exe is no valueble for nvim

File: cmake/FindJeMalloc.cmake (hunk 1 above)
Comment (justinmk) [reply cmake/FindJeMalloc.cmake]: But we should _never_ use jemalloc on Windows. It serves no purpose. So `if(WIN32)` should instead disable jemalloc.

File: cmake/FindUnibilium.cmake (hunk 2 above)
Comment (justinmk) [reply cmake/FindUnibilium.cmake]: All of this static stuff should be controlled by a single global option, `NVIM_BUILD_STATIC`. 

Instead of checking `if(WIN32)`, check `if(NVIM_BUILD_STATIC)`.

Comment (zhaozg) [reply cmake/FindUnibilium.cmake]: Like NVIM_BUILD_STATIC, give choice right to dev/users. I'll add that.

File: cmake/FindJeMalloc.cmake (hunk 1 above)
Comment (zhaozg) [reply cmake/FindJeMalloc.cmake]: After think, I hold point that as a way of performance optimization, we should enable on windows if found jemalloc already exist.

Comment (justinmk) [reply cmake/FindJeMalloc.cmake]: No, jemalloc should never be used on Windows. The Windows allocator does a better job than jemalloc.

Comment (zhaozg) [reply cmake/FindJeMalloc.cmake]: ok

File: CMakeLists.txt [hunk 4]
```diff
@@ -407,16 +412,16 @@ if (ENABLE_JEMALLOC)
...
-
 find_package(Iconv)
 if(Iconv_FOUND)
   include_directories(SYSTEM ${Iconv_INCLUDE_DIRS})
 endif()
 
+find_package(LibIntl)
+if(LibIntl_FOUND)
+  include_directories(SYSTEM ${LibIntl_INCLUDE_DIRS})
+endif()
```
Comment (justinmk) [inline CMakeLists.txt]: Was this moved below the Icon stuff for a reason?

File: src/nvim/CMakeLists.txt [hunk 5]
```diff
@@ -401,6 +406,12 @@ if(JEMALLOC_FOUND)
   list(APPEND NVIM_EXEC_LINK_LIBRARIES ${JEMALLOC_LIBRARIES})
 endif()
 
+if(WIN32 AND NVIM_BUILD_STATIC)
+  list(REMOVE_DUPLICATES NVIM_EXEC_LINK_LIBRARIES)
+  # Static stdc++ runtime library must keep tail for static link 
+  list(APPEND NVIM_EXEC_LINK_LIBRARIES -static-libgcc -static-libstdc++ -Wl,-Bstatic -lstdc++)
```
Comment (justinmk) [inline src/nvim/CMakeLists.txt]: Maybe it's worth doing REMOVE_DUPLICATES after this line, then append `-lstdc++`. Like this:

```
list(APPEND NVIM_EXEC_LINK_LIBRARIES -static-libgcc -static-libstdc++ -Wl,-Bstatic)
list(REMOVE_DUPLICATES NVIM_EXEC_LINK_LIBRARIES)
list(APPEND NVIM_EXEC_LINK_LIBRARIES -lstdc++)
```

Also, should there be a space in `-Wl,-Bstatic`? Like this: `-Wl, -Bstatic`

File: third-party/CMakeLists.txt [hunk 6]
```diff
@@ -103,29 +104,20 @@ set(LUA_SHA256 2640fc56a795f29d28ef15e13c34a47e223960b0240e8cb0a82d9b0738695333)
...
-else()
-set(LIBTERMKEY_URL http://www.leonerd.org.uk/code/libtermkey/libtermkey-0.20.tar.gz)
-set(LIBTERMKEY_SHA256 6c0d87c94ab9915e76ecd313baec08dedf3bd56de83743d9aa923a081935d2f5)
-endif()
-
 set(LIBVTERM_URL https://github.com/neovim/libvterm/archive/a9c7c6fd20fa35e0ad3e0e98901ca12dfca9c25c.tar.gz)
 set(LIBVTERM_SHA256 1a4272be91d9614dc183a503786df83b6584e4afaab7feaaa5409f841afbd796)
 
-set(JEMALLOC_URL https://github.com/jemalloc/jemalloc/releases/download/4.5.0/jemalloc-4.5.0.tar.bz2)
-set(JEMALLOC_SHA256 9409d85664b4f135b77518b0b118c549009dc10f6cba14557d170476611f6780)
```
Comment (justinmk) [inline third-party/CMakeLists.txt]: This can be left as it was.

File: third-party/CMakeLists.txt [hunk 7]
```diff
@@ -103,29 +104,20 @@ set(LUA_SHA256 2640fc56a795f29d28ef15e13c34a47e223960b0240e8cb0a82d9b0738695333)
 set(LUAROCKS_URL https://github.com/luarocks/luarocks/archive/v2.4.3.tar.gz)
 set(LUAROCKS_SHA256 ea1881d6954f2a98c34f93674571c8f0cbdbc28dedb3fa3cb56b6a91886d1a99)
 
-set(UNIBILIUM_URL https://github.com/mauke/unibilium/archive/v1.2.1.tar.gz)
-set(UNIBILIUM_SHA256 6045b4f6adca7b1123284007675ca71f718f70942d3a93d8b9fa5bd442006ec1)
```
Comment (justinmk) [inline third-party/CMakeLists.txt]: This can be left as it was.

File: third-party/CMakeLists.txt [hunk 8]
```diff
@@ -103,29 +104,20 @@ set(LUA_SHA256 2640fc56a795f29d28ef15e13c34a47e223960b0240e8cb0a82d9b0738695333)
...
-
-if(WIN32)
-set(LIBTERMKEY_URL https://github.com/equalsraf/libtermkey/archive/tb-windows.zip)
-set(LIBTERMKEY_SHA256 c81e33e38662b151a49847ff4feef4f8c4b2a66f3e159a28b575cbc9bcd8ffea)
-else()
-set(LIBTERMKEY_URL http://www.leonerd.org.uk/code/libtermkey/libtermkey-0.20.tar.gz)
-set(LIBTERMKEY_SHA256 6c0d87c94ab9915e76ecd313baec08dedf3bd56de83743d9aa923a081935d2f5)
-endif()
```
Comment (justinmk) [inline third-party/CMakeLists.txt]: IIUC this doesn't need to change. Let's try to minimize the changes.

File: src/nvim/CMakeLists.txt (hunk 5 above)
Comment (zhaozg) [reply src/nvim/CMakeLists.txt]: > Also, should there be a space in -Wl,-Bstatic? Like this: -Wl, -Bstatic (I don't know, just asking)

According to ld manual,
```
---if the linker is being invoked indirectly, via a compiler driver (e.g. gcc) then all the linker command line options should be prefixed by -Wl,(or whatever is appropriate for the particular compiler driver) like this:

gcc -Wl,--start-group foo.o bar.o -Wl,--end-group
```
So it should not separate by a space.

Comment (zhaozg) [reply src/nvim/CMakeLists.txt]: Done, followed you advice, but minor change.

**Comment by zhaozg** [comment]
All comments have been processed.

File: CMakeLists.txt (hunk 4 above)
Comment (justinmk) [inline CMakeLists.txt]: Why was this moved?

File: cmake/FindUnibilium.cmake [hunk 9]
```diff
@@ -23,6 +23,10 @@ find_path(UNIBILIUM_INCLUDE_DIR unibilium.h
           PATHS ${PC_UNIBILIUM_INCLUDEDIR} ${PC_UNIBILIUM_INCLUDE_DIRS}
           ${LIMIT_SEARCH})
 
+if(NVIM_BUILD_STATIC OR WIN32)
```
Comment (justinmk) [inline cmake/FindUnibilium.cmake]: Why is "OR WIN32" still here?  Should be controlled by NVIM_BUILD_STATIC.

Comment (zhaozg) [reply cmake/FindUnibilium.cmake]: Because Mingw64 only have static UNIBILIUM libs, never dynamics.

File: CMakeLists.txt (hunk 4 above)
Comment (zhaozg) [reply CMakeLists.txt]: Because intl depends on iconv, findlibintl.cmake will use libIconv, so do this.

File: cmake/FindUnibilium.cmake (hunk 9 above)
Comment (zhaozg) [reply cmake/FindUnibilium.cmake]: image

**Comment by zhaozg** [comment]
OK, night here, I have to bed, and If comment, I'll fix tomorrow. thanks.

**Comment by zhaozg** [comment]
ping @justinmk

**Comment by zhaozg** [comment]
[Download static neovim will dynamic winpty](https://ci.appveyor.com/project/neovim/neovim/build/18360/job/f3sm3yifq4tyx8yo/artifacts)

**Comment by justinmk** [comment]
Hmm, the size of the static artifacts is bigger:

```
build\bin\nvim.exe 22 MB
build\Neovim.zip 27 MB
 ```

non-static:

```
build\bin\nvim.exe 19 MB
build\Neovim.zip 25 MB 
```

File: CMakeLists.txt [hunk 10]
```diff
@@ -396,6 +396,11 @@ else()
   option(ENABLE_JEMALLOC "enable jemalloc" ON)
 endif()
 
+if(CMAKE_SYSTEM_NAME MATCHES "Windows")
+  message(STATUS "detected Windows; disabled jemalloc.")
+  option(ENABLE_JEMALLOC "enable jemalloc" OFF)
+endif()
```
Comment (justinmk) [inline CMakeLists.txt]: Why is this needed? `BuildJeMalloc.cmake` already skips itself if WIN32 is detected.

**Comment by zhaozg** [comment]
after strip will small
```
2018/02/03  10:46    <DIR>          .               
2018/02/03  10:46    <DIR>          ..              
2018/01/25  17:45           261,889 ca-bundle.crt   
2018/01/25  17:45            16,384 cat.exe         
2018/01/25  17:45         1,694,208 curl.exe        
2018/01/25  17:45            68,096 diff.exe        
2018/01/29  09:37        38,964,929 nvim-qt.exe     
2018/02/12  15:50         7,860,512 nvim.exe        
2018/01/25  17:45            59,392 tee.exe         
2018/01/25  17:45           483,840 tidy.exe        
2018/01/25  17:45           912,384 win32yank.exe   
2018/01/25  17:45         1,221,101 winpty-agent.exe
              10 个文件     51,542,735 字节              
```
but nvim-qt.exe static link keep huge after strip

**Comment by justinmk** [comment]
@zhaozg Thank you very much for your effort here. However, at present my impression is that it makes the build more complicated without having much benefit. For example the need for `COMPILING_WINPTY_DLL` is an extra complication.

Also this would need more work after the recent changes which added support for MSVC.
